File: src/LanAudioRelay.Core/Audio/JitterBuffer.cs
using LanAudioRelay.Core.Protocol;

namespace LanAudioRelay.Core.Audio;

public sealed class JitterBuffer
{
    private readonly object _gate = new();
    private readonly SortedDictionary<uint, AudioPacket> _frames = new();
    private readonly int _targetDepth;
    private readonly int _maxFrames;
    private bool _started;
    private uint _expectedSequence;

    public JitterBuffer(int targetDepth = 3, int maxFrames = 128)
    {
        if (targetDepth < 1)
        {
            throw new ArgumentOutOfRangeException(nameof(targetDepth));
        }

        if (maxFrames <= targetDepth)
        {
            throw new ArgumentOutOfRangeException(nameof(maxFrames));
        }

        _targetDepth = targetDepth;
        _maxFrames = maxFrames;
    }

    public int BufferedFrameCount
    {
        get
        {
            lock (_gate)
            {
                return _frames.Count;
            }
        }
    }

    public long MissingFrames { get; private set; }

    public void Reset()
    {
        lock (_gate)
        {
            _frames.Clear();
            _started = false;
            _expectedSequence = 0;
            MissingFrames = 0;
        }
    }

    public void Push(AudioPacket packet)
    {
        lock (_gate)
        {
            if (_started && packet.Sequence < _expectedSequence)
            {
                return;
            }

            _frames[packet.Sequence] = packet;

            while (_frames.Count > _maxFrames)
            {
                _frames.Remove(_frames.Keys.First());
            }
        }
    }

    public bool TryPop(out AudioPacket? packet)
    {
        lock (_gate)
        {
            packet = null;

            if (!_started)
            {
                if (_frames.Count < _targetDepth)
                {
                    return false;
                }

                _expectedSequence = _frames.Keys.First();
                _started = true;
            }

            if (_frames.Remove(_expectedSequence, out packet))
            {
                _expectedSequence++;
                return true;
            }

            if (_frames.Count == 0)
            {
                _started = false;
                return false;
            }

            MissingFrames++;
            _expectedSequence++;
            return false;
        }
    }
}

File: mac/LanAudioRelayMac/Sources/LanAudioRelayMacCore/JitterBuffer.swift
import Foundation

public final class JitterBuffer: @unchecked Sendable {
    private let lock = NSLock()
    private var frames: [UInt32: AudioPacket] = [:]
    private let targetDepth: Int
    private let maxFrames: Int
    private var started = false
    private var expectedSequence: UInt32 = 0

    public private(set) var missingFrames: Int64 = 0

    public init(targetDepth: Int = 3, maxFrames: Int = 128) {
        precondition(targetDepth >= 1)
        precondition(maxFrames > targetDepth)
        self.targetDepth = targetDepth
        self.maxFrames = maxFrames
    }

    public var bufferedFrameCount: Int {
        lock.withLock { frames.count }
    }

    public func reset() {
        lock.withLock {
            frames.removeAll()
            started = false
            expectedSequence = 0
            missingFrames = 0
        }
    }

    public func push(_ packet: AudioPacket) {
        lock.withLock {
            if started && packet.sequence < expectedSequence {
                return
            }

            frames[packet.sequence] = packet

            while frames.count > maxFrames, let first = frames.keys.min() {
                frames.removeValue(forKey: first)
            }
        }
    }

    public func pop() -> AudioPacket? {
        lock.withLock {
            if !started {
                guard frames.count >= targetDepth, let first = frames.keys.min() else {
                    return nil
                }
                expectedSequence = first
                started = true
            }

            if let packet = frames.removeValue(forKey: expectedSequence) {
                expectedSequence &+= 1
                return packet
            }

            if frames.isEmpty {
                started = false
                return nil
            }

            missingFrames += 1
            expectedSequence &+= 1
            return nil
        }
    }
}

private extension NSLock {
    func withLock<T>(_ body: () throws -> T) rethrows -> T {
        lock()
        defer { unlock() }
        return try body()
    }
}
